File: src/db/read.rs
use super::commit::{Commit, FromHashError};
use super::index;
use crate::dag;
use crate::prolly;
use std::collections::hash_map::HashMap;
use std::convert::TryInto;

#[derive(Debug)]
pub enum Whence {
    Head(String),
    #[allow(dead_code)]
    Hash(String),
}

pub struct OwnedRead<'a> {
    dag_read: dag::OwnedRead<'a>,
    map: prolly::Map,
    indexes: HashMap<String, index::Index>,
}

#[derive(Debug)]
pub enum ReadCommitError {
    CommitFromHeadError(FromHashError),
    GetHeadError(dag::Error),
    MapLoadError(prolly::LoadError),
    UnknownHead(String),
}

impl<'a> OwnedRead<'a> {
    pub async fn from_whence(
        whence: Whence,
        dag_read: dag::OwnedRead<'a>,
    ) -> Result<OwnedRead<'a>, ReadCommitError> {
        let (_, basis, map) = read_commit(whence, &dag_read.read()).await?;
        let indexes = read_indexes(&basis);
        Ok(OwnedRead {
            dag_read,
            map,
            indexes,
        })
    }

    pub fn as_read(&'a self) -> Read<'a> {
        Read::new(self.dag_read.read(), &self.map, &self.indexes)
    }
}

pub async fn read_commit(
    whence: Whence,
    read: &dag::Read<'_>,
) -> Result<(String, Commit, prolly::Map), ReadCommitError> {
    use ReadCommitError::*;
    let hash = match whence {
        Whence::Hash(s) => s,
        Whence::Head(s) => read
            .get_head(&s)
            .await
            .map_err(GetHeadError)?
            .ok_or_else(|| UnknownHead(s.to_string()))?,
    };
    let commit = Commit::from_hash(&hash, read)
        .await
        .map_err(CommitFromHeadError)?;
    let map = prolly::Map::load(commit.value_hash(), read)
        .await
        .map_err(MapLoadError)?;
    Ok((hash, commit, map))
}

pub fn read_indexes(commit: &Commit) -> HashMap<String, index::Index> {
    commit
        .indexes()
        .iter()
        .map(|meta| {
            (
                meta.definition.name.clone(),
                index::Index::new(meta.clone(), None),
            )
        })
        .collect()
}

pub struct Read<'a> {
    #[allow(dead_code)]
    dag_read: dag::Read<'a>,
    map: &'a prolly::Map,
    indexes: &'a HashMap<String, index::Index>,
}

impl<'a> Read<'a> {
    pub fn new(
        dag_read: dag::Read<'a>,
        map: &'a prolly::Map,
        indexes: &'a HashMap<String, index::Index>,
    ) -> Read<'a> {
        Read {
            dag_read,
            map,
            indexes,
        }
    }

    pub fn has(&self, key: &[u8]) -> bool {
        self.map.has(key)
    }

    pub fn get(&self, key: &[u8]) -> Option<&[u8]> {
        self.map.get(key)
    }

    pub async fn scan(
        &'a self,
        opts: super::ScanOptions,
        callback: impl Fn(super::scan::ScanResult<'_>),
    ) -> Result<(), ScanError> {
        use ScanError::*;

        let opts_internal: super::scan::ScanOptionsInternal =
            opts.try_into().map_err(ScanError::ScanOptionsError)?;

        match &opts_internal.index_name {
            Some(name) => {
                let idx = self
                    .indexes
                    .get(name)
                    .ok_or_else(|| UnknownIndexName(name.to_string()))?;
                let guard = idx.get_map(&self.dag_read).await.map_err(GetMapError)?;
                super::scan::scan(guard.get_map(), opts_internal).for_each(callback)
            }
            None => super::scan::scan(&self.map, opts_internal).for_each(callback),
        };
        Ok(())
    }
}

#[derive(Debug)]
pub enum ScanError {
    GetMapError(index::GetMapError),
    ScanOptionsError(super::scan::ScanOptionsError),
    UnknownIndexName(String),
}

#[cfg(test)]
mod tests {
    use super::super::*;
    use super::*;
    use crate::dag;
    use crate::db;
    use crate::db::write::init_db;
    use crate::kv::memstore::MemStore;
    use crate::util::rlog::LogContext;
    use str_macro::str;

    #[async_std::test]
    async fn basics() {
        let ds = dag::Store::new(Box::new(MemStore::new()));
        init_db(
            ds.write(LogContext::new()).await.unwrap(),
            db::DEFAULT_HEAD_NAME,
        )
        .await
        .unwrap();
        let mut w = write::Write::new_local(
            Whence::Head(str!(db::DEFAULT_HEAD_NAME)),
            str!("mutator_name"),
            serde_json::Value::Array(vec![]).to_string(),
            None,
            ds.write(LogContext::new()).await.unwrap(),
        )
        .await
        .unwrap();
        w.put(
            LogContext::new(),
            "foo".as_bytes().to_vec(),
            "bar".as_bytes().to_vec(),
        )
        .await
        .unwrap();
        w.commit(db::DEFAULT_HEAD_NAME).await.unwrap();

        let dr = ds.read(LogContext::new()).await.unwrap();
        let r = OwnedRead::from_whence(Whence::Head(str!(db::DEFAULT_HEAD_NAME)), dr)
            .await
            .unwrap();
        let rr = r.as_read();
        let val = rr.get("foo".as_bytes());
        assert_eq!(Some("bar".as_bytes()), val);
    }
}

File: src/db/read.ts
import {IndexRead} from './index';
import type * as dag from '../dag/mod';
import {
  Commit,
  DEFAULT_HEAD_NAME,
  fromHash as commitFromHash,
  Meta,
} from './commit';
import type {ReadonlyJSONValue} from '../json';
import {BTreeRead, BTreeWrite} from '../btree/mod';
import type {Hash} from '../hash';

export class Read {
  private readonly _dagRead: dag.Read;
  map: BTreeRead;
  readonly indexes: Map<string, IndexRead>;
  shouldDeepClone = false;

  constructor(
    dagRead: dag.Read,
    map: BTreeRead,
    indexes: Map<string, IndexRead>,
  ) {
    this._dagRead = dagRead;
    this.map = map;
    this.indexes = indexes;
  }

  has(key: string): Promise<boolean> {
    return this.map.has(key);
  }

  get(key: string): Promise<ReadonlyJSONValue | undefined> {
    return this.map.get(key);
  }

  isEmpty(): Promise<boolean> {
    return this.map.isEmpty();
  }

  async getMapForIndex(indexName: string): Promise<BTreeRead> {
    const idx = this.indexes.get(indexName);
    if (idx === undefined) {
      throw new Error(`Unknown index name: ${indexName}`);
    }
    return idx.withMap(this._dagRead, map => map);
  }

  get closed(): boolean {
    return this._dagRead.closed;
  }

  close(): void {
    this._dagRead.close();
  }
}

const enum WhenceType {
  Head,
  Hash,
}

export type Whence =
  | {
      type: WhenceType.Hash;
      hash: Hash;
    }
  | {
      type: WhenceType.Head;
      name: string;
    };

export function whenceHead(name: string): Whence {
  return {
    type: WhenceType.Head,
    name,
  };
}

export function whenceHash(hash: Hash): Whence {
  return {
    type: WhenceType.Hash,
    hash,
  };
}

export function readFromDefaultHead(dagRead: dag.Read): Promise<Read> {
  return fromWhence(whenceHead(DEFAULT_HEAD_NAME), dagRead);
}

export async function fromWhence(
  whence: Whence,
  dagRead: dag.Read,
): Promise<Read> {
  const [, basis, map] = await readCommitForBTreeRead(whence, dagRead);
  const indexes = readIndexesForRead(basis);
  return new Read(dagRead, map, indexes);
}

export async function readCommit(
  whence: Whence,
  dagRead: dag.Read,
): Promise<[Hash, Commit<Meta>]> {
  let hash: Hash;
  switch (whence.type) {
    case WhenceType.Hash:
      hash = whence.hash;
      break;
    case WhenceType.Head: {
      const h = await dagRead.getHead(whence.name);
      if (h === undefined) {
        throw new Error(`Unknown head: ${whence.name}`);
      }
      hash = h;
      break;
    }
  }

  const commit = await commitFromHash(hash, dagRead);
  return [hash, commit];
}

export async function readCommitForBTreeRead(
  whence: Whence,
  dagRead: dag.Read,
): Promise<[Hash, Commit<Meta>, BTreeRead]> {
  const [hash, commit] = await readCommit(whence, dagRead);
  return [hash, commit, new BTreeRead(dagRead, commit.valueHash)];
}

export async function readCommitForBTreeWrite(
  whence: Whence,
  dagWrite: dag.Write,
): Promise<[Hash, Commit<Meta>, BTreeWrite]> {
  const [hash, commit] = await readCommit(whence, dagWrite);
  return [hash, commit, new BTreeWrite(dagWrite, commit.valueHash)];
}

export function readIndexesForRead(
  commit: Commit<Meta>,
): Map<string, IndexRead> {
  const m = new Map();
  for (const index of commit.indexes) {
    m.set(index.definition.name, new IndexRead(index, undefined));
  }
  return m;
}
